doc/builders/fetchers: Add examples, reduce sentence complexity (#179489)
NixOS/nixpkgs: doc/builders/fetchers.chapter.md

# Fetchers {#chap-pkgs-fetchers}

> Nixpkgs provides fetchers for different protocols and services. Fetchers are functions that simplify downloading files.

Reviewer: ```suggestion
Building software with Nix often requires downloading source code and other files from the internet.
`nixpkgs` provides fetchers for different protocols and services. Fetchers are functions that simplify downloading files.
```

> Nixpkgs provides fetchers for different protocols and services. Fetcher's are functions that simplify downloading files.

Reviewer 2: This is the introduction of the fetchers chapter. While the previous intro certainly wasn't great, you've removed essential information from it. It should answer succinctly:

- What is a fetcher?
- What is the purpose of a fetcher?

Furthermore, please fix a pervasive spelling mistake.

```diff
-fetcher's
+fetchers
```

"fetcher's" has a possessive meaning. Example: a fetcher's hash makes sure [...]
The plural is just "fetchers".

Author: Good catch on the extra apostrophes. Updated.

> Fetchers are functions that simplify downloading files.

What is missing from this? Is it too simplistic?

## Caveats

> Fetchers create [_fixed output derivations_](https://nixos.org/manual/nix/stable/#fixed-output-drvs) from downloaded files. Later, Nix can reuse the downloaded files via their hash. While the caching improves performance, it can lead to some confusion. For example, consider the following fetcher:

Reviewer 2: ```suggestion
Fetchers create [_fixed output derivations_](https://nixos.org/manual/nix/stable/#fixed-output-drvs) that downloaded files. Later, Nix can reuse the downloaded files via their hash. While the caching improves performance, it can lead to some confusion. For example, consider the following fetcher:
```

> Fetchers create [_fixed output derivations_](https://nixos.org/manual/nix/stable/#fixed-output-drvs) that downloaded files. Later, Nix can reuse the downloaded files via their hash. While the caching improves performance, it can lead to some confusion. For example, consider the following fetcher:

Reviewer 3: “that downloaded files” sounds weird as well.

“reuse the downloaded files via their hash” this has nothing to do with the hash of downloaded files – the hash is of the output plus some other attributes passed to the derivation.

Reviewer: ```suggestion
Fetchers create [fixed output derivations](https://nixos.org/manual/nix/stable/#fixed-output-drvs) from downloaded files.
Nix can reuse the downloaded files via the hash of the resulting derivation.

The fact that the hash belongs to the Nix derivation output and not the file itself can lead to confusion.
For example, consider the following fetcher:
```

> Fetcher's store their output in the nix store, or cache, using Nix's [_fixed output derivation_](https://nixos.org/manual/nix/stable/#fixed-output-drvs). Later, Nix can reuse the downloaded contents via their hash. While the caching improves performance, it can lead to some confusion. For example, consider the following fetcher:

Reviewer 3: “Fetcher's store their output in the nix store, or cache” gives agency to fetchers. And while it would be sort of close enough when talking about writing into Nix store – though, to be precise, the fetchers are just helpers to create fixed-output derivations (download & build instructions) and Nix is responsible for their realization – they definitely have nothing to do with cache.

Author: What about:

> Fetchers create fixed output derivations from downloaded files. Later, Nix can reuse the downloaded files via their hash.

```nix
fetchurl {
  url = "http://www.example.org/hello-1.0.tar.gz";
  sha256 = "1111111111111111111111111111111111111111111111111111";
};
```

A common mistake is to update a fetcher’s URL, or a version parameter, without updating the hash.

```nix
fetchurl {
  url = "http://www.example.org/hello-1.1.tar.gz";
  sha256 = "1111111111111111111111111111111111111111111111111111";
};

> ```

Reviewer: Having all `1` as a hash may be confusing, because it does not really look like a typical hash. What about using a string that looks more like a hash?

Author: Updated to a random example sha.

> **This will reuse the old contents**. Remember to invalidate the hash argument by setting the sha256 to an empty string.

Reviewer: ```suggestion
**This will reuse the old contents**.
Remember to invalidate the hash argument, in this case by setting the `sha256` attribute to an empty string.
```

```nix
fetchurl {
  url = "http://www.example.org/hello-1.1.tar.gz";

>   sha256 = "0000000000000000000000000000000000000000000000000000";

Reviewer 4: since nix 2.4 this can just be empty to achieve the same effect

https://github.com/nixos/nix/commit/762273f1fd48b2b2f2bbedca65abfd4c07b5af05

Author: Rad, updated!

};
```

> Hash mismatches generate an error message containing the correct sha256.

Reviewer: ```suggestion
Use the resulting error message to determine the correct hash.

<error message here>
```

[NixOS Wiki as an old article on this](https://nixos.wiki/wiki/FAQ/Pinning_Nixpkgs) (also [this one](https://nixos.wiki/wiki/How_to_fetch_Nixpkgs_with_an_empty_NIX_PATH)), and this is probably where most people learn the `nix-prefetch-url` trick. Not sure if we should include it here as well, as it is obviously less convenient than just taking the result of the error message.

> A similar problem arises while testing changes to a fetcher's implementation. If the output of the derivation already exists in the nix store or cache, test failures can go undetected. The [`invalidateFetcherByDrvHash`](#tester-invalidateFetcherByDrvHash) function helps prevent reusing cached derivations.

Reviewer: ```suggestion
A similar problem arises while testing changes to a fetcher's implementation. If the output of the derivation already exists in the Nix store, test failures can go undetected. The [`invalidateFetcherByDrvHash`](#tester-invalidateFetcherByDrvHash) function helps prevent reusing cached derivations.
```

> A similar problem arises when changing a fetcher's implementation. Changes may cause a fixed output derivation to fail. Failures can go undetected if the output of the derivation already exists in the nix store or cache. The [`invalidateFetcherByDrvHash`](#tester-invalidateFetcherByDrvHash) function helps prevent this. It uses the derivation hash in the output's name to invalidate the cache.

Reviewer 2: This last sentence is also missing important context. Admittedly this wasn't very clear in the current version either, but we're talking about running fixed output derivations for the purpose of testing, why that is tricky and what the solution is.

Reviewer 3: I think talking about cache invalidation is confusing, since the store is, from the viewpoint of the most common operations, append-only. Cache invalidation IMO implies that an entry matching a certain key is marked invalid, but here the function just changes the look-up key.

Reviewer 2: You're right. It would be more accurate to change "invalidate" to "defeat", or to describe the solution in more literal terms.

Author: What about:

```
A similar problem arises while testing changes to a fetcher's implementation. If the output of the derivation already exists in the nix store or cache, test failures can go undetected. The invalidateFetcherByDrvHash function helps prevent reusing cached derivations. 
```

Readers can follow the link to dive into the function's details.
